phoenix-core-client/src/main/java/org/apache/phoenix/jdbc/PhoenixConnection.java

```diff
@@ -761,13 +761,22 @@ public PTable getTable(String name) throws SQLException {
     }
 
     public PTable getTable(PTableKey key) throws SQLException {
-        PTable table;
-        try {
-            table = getTableRef(key).getTable();
-        } catch (TableNotFoundException e) {
-            table =  getTableNoCache(key.getName());
+        PTableRef tableRef = getTableRefNoTNFE(key);
+        if (tableRef == null && key.getTenantId() != null) {
+            tableRef = getTableRefNoTNFE(new PTableKey(null, key.getName()));
```

Comment:
Consider the case when there is a tenant view on the table and the tenant has the same name as the underlying table. I think Phoenix allows that. Then if the base table is in the cache but the tenant view is not in the cache you will return the base table object.

Reply:
Then I think what we need to do address PHOENIX-4475. I can take up PHOENIX-4475 and address both PHOENIX-4475 and PHOENIX-7484 in one PR. If we don't use UCF and always hit SYSCAT then its a huge regression with tenant connection compared to non-tenant connection.

Reply:
Another idea could be to prevent creating tenant views with same name as base tables?

Reply:
> Then if the base table is in the cache but the tenant view is not in the cache you will return the base table object.

This will happen now also with MetaDataClient#updateCache() but it relies on UCF to get cache refreshed eventually as per my understanding. Please correct me if I am wrong.

Reply:
> Another idea could be to prevent creating tenant views with same name as base tables?
You can introduce a name conflict when creating a tenant view with another non related table also. It doesn't even need to be the base table.

Reply:
> You can introduce a name conflict when creating a tenant view with another non related table also. It doesn't even need to be the base table.

Got it. So how about: [this one](https://github.com/apache/phoenix/pull/2041#discussion_r1887104412). I think this will work, right?

Reply:
I also think it as an application problem. You are using a tenant connection but the entity you are writing to is a global object. Maybe you should be using a global connection in the first place.

Reply:
> Maybe you should be using a global connection in the first place.

With global connection I feel one tenant can modify the records of other tenants by providing some other tenant Id in UPSERT. With tenant specific connection I think we able to impose tenant boundaries and prevent one tenant from modifying records of other tenants. I think on a multi-tenant base table we should be able to read and write using tenant connection. WDYT?

Reply:
Well, that's a weak argument because if I can specify some other tenantid in UPSERT, what's stopping to create a tenant connection with someone else's tenantid.

Reply:
Yeah, that's true. For queries we use tenant connection to query a multi-tenant base table but for writes we need to use global connection? There is nothing wrong in it as such but does seem odd to me.

Reply:
> For queries we use tenant connection to query a multi-tenant base table 

Where is that mandated ? 

You need to think in terms of the entity you are reading and writing to. If the entity is a TENANT VIEW  use a tenant connection otherwise use a global connection.